### DISCUSSION TITLE
Async open nvim-tree on startup?

### DISCUSSION THREAD

**Discussion by linrongbin16** [nvim-tree/nvim-tree.lua#2002] [Ideas]
Hi, I'm using nvim-tree with lazy.nvim, and my config open nvim-tree on startup for a directory.

Here's my 1st-edition auto-open nvim-tree on startup augroup:
```lua
vim.api.nvim_create_augroup("nvim_tree_augroup", { clear = true })
vim.api.nvim_create_autocmd("VimEnter", {
    group = "nvim_tree_augroup",
    callback = function(data)
        -- buffer is a [No Name]
        local no_name = data.file == "" and vim.bo[data.buf].buftype == ""
        -- buffer is a directory
        local directory = vim.fn.isdirectory(data.file) == 1
        if not no_name and not directory then
            return
        end

        -- change to the directory
        if directory then
            vim.cmd.cd(data.file)
        end
        -- open the tree
        require("nvim-tree.api").tree.open()
    end,
})
```

And in lazy's profile, I found `nvim_tree_augroup` cost ~30ms.
<img width="1137" alt="image" src="https://user-images.githubusercontent.com/6496887/219318745-d817a845-4618-4c94-abf9-3caed852020c.png">
<img width="621" alt="image" src="https://user-images.githubusercontent.com/6496887/219318900-acf20ecd-53b3-43f2-bcb9-8b495a94c3bf.png">

Guess the most heavy part is `cd` directory and `nvim-tree.api.open` API call.

So I use `vim.defer_fn` API to wrap this part(with 0ms delay), here's the 2nd-edition:
```lua
vim.api.nvim_create_augroup("nvim_tree_augroup", { clear = true })
vim.api.nvim_create_autocmd("VimEnter", {
    group = "nvim_tree_augroup",
    callback = function(data)
        -- use defer_fn to open async
        local function open_impl()
            -- buffer is a [No Name]
            local no_name = data.file == "" and vim.bo[data.buf].buftype == ""
            -- buffer is a directory
            local directory = vim.fn.isdirectory(data.file) == 1
            if not no_name and not directory then
                return
            end

            -- change to the directory
            if directory then
                vim.cmd.cd(data.file)
            end
            -- open the tree
            require("nvim-tree.api").tree.open()
        end
        vim.defer_fn(open_impl, 0)
    end,
})
```
Here's the lazy's profile:
<img width="1146" alt="image" src="https://user-images.githubusercontent.com/6496887/219319172-68c68c17-178a-4ffa-ad6d-c039c7335af6.png">
<img width="696" alt="image" src="https://user-images.githubusercontent.com/6496887/219319280-f20ec2ef-2aec-4b96-b593-259b91f4570f.png">

The interesting thing is: total time is almost the same, `nvim_tree_augroup` time is reduced &lt; 0.1ms. My eyes feel that it's much faster and smooth.

I'm not sure why is that?

**Comment by alex-courtis** [comment]
The 30ms is not surprising: nvim-tree will enumerate the file system, git, diagnostics etc.

I'm not sure why that feels smoother; as you say the end to end time is the same. Perhaps the ordering changes around updating statusline, other windows/plugins etc. contribute to this.
